src/libs/OptionsListUtilsLHN.js

```diff
@@ -0,0 +1,312 @@
+import Onyx from 'react-native-onyx';
+import _ from 'underscore';
+import Str from 'expensify-common/lib/str';
+import ONYXKEYS from '../ONYXKEYS';
+import * as ReportUtils from './ReportUtils';
+import * as Localize from './Localize';
+import CONST from '../CONST';
+import * as OptionsListUtils from './OptionsListUtils';
+import * as CollectionUtils from './CollectionUtils';
+import Permissions from './Permissions';
+
+let reports;
+Onyx.connect({
+    key: ONYXKEYS.COLLECTION.REPORT,
+    waitForCollectionCallback: true,
+    callback: val => reports = val,
+});
```

Comment:
haha, maybe this is why all my unit tests are failing :ohnothing:

It works in the UI just fine, but I need to explore this more.

I really appreciate that breakdown, but it doesn't quite fit with how I think about it. Maybe you can spot where we are on a different wavelength.

1. Utility file connects to onyx `Onyx.connect({key: 'one'});`
2. Utility file uses that value in a method `getData()`
3. Component uses `withOnyx({one: {key: 'one'}})`
4. Component calls `getData()` from the render method

So far, that's just a simple setup for how this code is currently structured.

5. The component is not going to be rendered at all until the data is read from Onyx because of [this](https://github.com/Expensify/react-native-onyx/blob/main/lib/withOnyx.js#L159-L161). So we know that there is no race condition with the component having access to the Onyx data.
6. Once Onyx reads the data, `keyChanged()` is called and there is a loop which goes through each subscription here: https://github.com/Expensify/react-native-onyx/blob/main/lib/Onyx.js#L333
7. The `Onyx.connect()` mappings end up being first, because those are added immediately with the JS is ran, and the `withOnyx()` mappings are only added once the component mounts. (the likelihood of it happening in a different order would really surprise me, that would mean components are mounted before the JS code is first evaluated)
8. Now that Onyx has read the data, the `Onyx.connect()` is updated, and now `setState()` is called on the component
9. The component calls `getData()` and we know that the utility file has all the proper data

Now, I'm not saying that this is absolutely 100% the say things happen, this is just how I think they happen. Like I said, I'm seeing some behavior in the unit tests which are challenging this thought process and hopefully you can help me understand where I'm making an incorrect assumption.

______________
> It seems to make sense that if the data changes the component subscribing to those changes would pass the data to a pure function in this file. Instead we have - something causes the component to update and hopefully the data here is up to date.

The thing that I just can't really fully get on board with is how passing 24,000 object references (4000 reports * 6 onyx keys) is more performant than having 6 Onyx connections in a utility file. I'm also not sure what that does to memory (since I haven't ever looked at the memory aspect of performance) So, while I agree in theory with what you are saying, when we are trying to squeek out every last bit of performance, this is perhaps a sacrifice we must make?

Reply:
That's a great description of the journey the data takes and thanks for putting those detailed thoughts down. I think all of those assumptions are fairly testable so we can get objective about this and just know for sure before leaning too hard in either direction. Either a race condition can happen or the changes we are making here are fine?

>The thing that I just can't really fully get on board with is how passing 24,000 object references (4000 reports * 6 onyx keys) is more performant than having 6 Onyx connections in a utility file.

I think your instinct is probably right here, but a separate concern that I didn't consider and didn't mean to imply that it would be _more_ performant. The suggestion to use `withOnyx()` is just provided so that access of data is safe and guaranteed.

>So, while I agree in theory with what you are saying, when we are trying to squeek out every last bit of performance, this is perhaps a sacrifice we must make?

I'd like to understand just a _little_ more about what the sacrifice is exactly first before deciding if it's worthy 😄 Maybe it's nothing and maybe there is no race condition at all to worry about or the places where I observed it don't apply in this new case.

Reply:
> The suggestion to use withOnyx() is just provided so that access of data is safe and guaranteed.

Ah, got it!

> I'd like to understand just a little more about what the sacrifice is exactly first before deciding if it's worthy

Fair point! To be clear, the sacrifice for me would be implementing something like `Welcome.js` everywhere that an action file uses `Onyx.connect()`. It feels like that's some kind of workaround, when maybe, we should just do something to ensure that it'll never end up with a race condition.

I can definitely write some tests to prove my theory that there is no race-condition, but maybe it would be a pretty attractive idea to see if we could do something to ensure there would not be a race-condition and we don't have to rely on assumptions or things like the order in which JS is evaluated.

I'm curious if you have any ideas on what an approach like that might look like? Could we build something like what is in `Welcome.js` into Onyx? I need to dig into that further to come up with my own ideas. I know we both have a lot of important irons in the fire currently.

Reply:
>we should just do something to ensure that it'll never end up with a race condition

👍 

Need to dig in more, but I did a quick test that I think shows some `withOnyx()` calls can be evaluated before `Onyx.connect()` by logging out whether a mapping has a `withOnyxInstance` or not and looks like the answer is yes. Though not sure if that proves there is any race condition.

```
[Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:913 [Onyx Connection] react component connected
2web.development.js?d0e7:913 [Onyx Connection] react component connected
web.development.js?d0e7:913 [Onyx Connection] react component connected
web.development.js?d0e7:913 [Onyx Connection] react component connected
4web.development.js?d0e7:913 [Onyx Connection] react component connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:913 [Onyx Connection] react component connected
3web.development.js?d0e7:913 [Onyx Connection] react component connected
web.development.js?d0e7:915 [Onyx Connection] library connected
web.development.js?d0e7:915 [Onyx Connection] library connected
9web.development.js?d0e7:913 [Onyx Connection] react component connected
web.development.js?d0e7:913 [Onyx Connection] react component connected
15web.development.js?d0e7:913 [Onyx Connection] react component connected
web.development.js?d0e7:913 [Onyx Connection] react component connected
34web.development.js?d0e7:913 [Onyx Connection] react component connected
2web.development.js?d0e7:913 [Onyx Connection] react component connected
web.development.js?d0e7:913 [Onyx Connection] react component connected
```

Most of the early ones are from libraries so it seems safe to assume that most would run quickly and maybe even return data before any react components render.

Reply:
Looking at `Welcome.js` again and I'm starting to think that the race condition we've seen happen because we assumed some values that would be available (need to be set from the server or read from Onyx) are always available (and in that case we're just holding stuff wrong). 

It's a time sensitive operation as we need one method to run only _after_ all the values we are looking for have been read or set from the server.

I'm not sure if the situation is the same in this PR. So the important questions are: 

Do we care if all of these values are available right away for the first login?

I think probably not.

And if any are updated later via the server will the sidebar re-render and show the correct data?

This one feels more important  and I think that's kind of where my uneasiness is with encouraging reliance on a big list of `Onyx.connect()` in a library method like this.

Calculating options depends on x, y, z variables, but we are not necessarily re-calculating when those options change. So I think this can be broken down into two problems:

1. There is no built-in way to guarantee a value has been set from the server yet for any method that might rely on `Onyx.connect()` values to be non null (`Welcome.js` suffers from this)

3. Using `Action.someMethod()` where there is some reliance on `Onyx.connect()` set values is kind of like a veiled `Onyx.get()` and is an inherently non-reactive pattern. The method caller has no way to know that the values have changed or that things should be recalculated (unless it's used somewhere that will subscribe to the keys and then re-rerun the logic - which requires a separate subscription somewhere else).

>I'm curious if you have any ideas on what an approach like that might look like?

I think both problems could potentially be solved with something like an "observable" e.g. if we wanted the sidebar to recalculate it's options when any of the data it depends on the build them changes we could have something like:

```js
const sidebarReportIDs = new Observable((subscriber) => {
    Onyx.connectMultiple(keys, (values) => {
        // runs when any of the values change
        subscriber.next(getOrderedReportIDs(values));
    });
});
```

```js
sidebarReportIDs.subscribe({
    next(reportIDs) {
        this.setState({reportIDList});
    },
});
```

Reply:
OK, that's interesting. I did a _slightly_ different test, and added this code to `keyChanged()`:

```
    _underscore.default.each(callbackToStateMapping, (mapping) => {

        if (mapping.callback) { console.log('connection')}
        else if (mapping.withOnyxInstance) { console.log('withOnyx')}
    });
```

That way it would test my assumption for point 7. The result is:

![image](https://user-images.githubusercontent.com/1228807/190830351-2fced007-8565-452f-960e-d9776ab2145a.png)

All but one connection mapping is listed first in the mappings. Not sure what's up with that little fella (maybe it's a connection triggered from a mounted component) or where that leaves us.

Reply:
(sorry, my comment came in out of order) It was referring to that test you did with the connection.

Reply:
I think taking a step back here... really the only reason this works in the SidebarLinks and SidebarUtils is because they are both connecting to the same Onyx Keys.

If they connected to different Onyx keys, then I could totally understand your logic.

However, I do see the danger in that. If someone else adds another Onyx connection to SidebarUtils and doesn't also add it to SidebarLinks `withOnyx()` then the values will remain stale in SidebarUtils.

The observable idea is interesting, but for that route, it seems the same as just connecting to the 6 onyx keys in OptionRowLHN.

Would you like me to experiment with using `withOnyx()` in `OptionRowLHN` and remove the connections from `SidebarUtils`? I will take a before/after snapshot, not only of CPU, but also of memory so we can see what happens.

Reply:
>However, I do see the danger in that. If someone else adds another Onyx connection to SidebarUtils and doesn't also add it to SidebarLinks withOnyx() then the values will remain stale in SidebarUtils.

👍 

>The observable idea is interesting, but for that route, it seems the same as just connecting to the 6 onyx keys in OptionRowLHN.

I think it's very similar, but not quite the same. If we compute the values we need outside of the React lifecycle we can then decide to re-render or not. A React component subscribing to `withOnyx()` is not render optimized by default and always wants to render when data changes (even if the thing that changes is something we don't care about). Though in theory we could tell it to "not render" with `shouldComponentUpdate()` or `React.memo()` (which we've done in the past and it works but is pretty messy).

**current pattern**
```
withOnyx updates with new data 
  -> component re-renders with new data because we are subscribed in the component
  -> calculation is done on render and with duplicate subscriptions data 
  -> (though we could make it a pure function and pass in all the data we need from SidebarLinks)
  -> render of SideBarLinks happens regardless of whether the computed value has changed 
```

**observable thing**
```
component subscribes to the computed data
  -> result in the lib is recalculated when new data appears
  -> we calculate the options and choose to trigger a render (or not)
  -> component only renders when the lib says it should
```

I don't know if this is actually something we totally need for every situation, but interesting to think about how to enable and build components that are much more selective about how they render. I think what you've got here should work great.

----
>Would you like me to experiment with using withOnyx() in OptionRowLHN and remove the connections from SidebarUtils

Could be interesting! I would probably expect to see a spike in those resources since we are going to add many copies of data to each sidebar list item and also possibly add extra renders since they could update when `SidebarLinks` updates and when any data they are subscribing to updates. With the changes now we've got two copies of the data (one in `SidebarLinks` and one in `SidebarUtils`), but the smaller `OptionRows` have a very small footprint as they are not subscribed to anything.

Reply:
OK, well, I'm at a loss for how to proceed then.

I think, for now, let's merge it as-is since everything works and it's proven to be performant. If problems arise, then we solve those problems later.

After sitting on this over the weekend, I don't necessarily think this PR is the one where we should experiment with different theories or patterns (like observable).

How's that sound?

Reply:
I also had one additional thought to help with the race-condition issue (but I don't really think it's a proven issue). In `keyChanged()`, we can first loop through all mappings that have callbacks. Once those callbacks are processed, then we do a second loop through all the mappings for `withOnyx()`. In that way, we will ensure that any `Onyx.connect()` mappings are updated before any component rerenders with new props.

Reply:
>I don't necessarily think this PR is the one where we should experiment with different theories or patterns (like observable)

I agree 💯 mainly just riffing in this thread.

>we can first loop through all mappings that have callbacks. Once those callbacks are processed, then we do a second loop through all the mappings for withOnyx(). In that way, we will ensure that any Onyx.connect() mappings are updated before any component rerenders with new props.

Definitely a creative solution. It feels like that would imply that standalone `connect()` calls should have some kind of implicit priority over React component calls. And I think the only reason you'd find that useful is if you use a pattern where there are duplicate subscriptions in both. Which might mean there's a better design we can reach for (but that's just my desire to have a single subscription or consistent way of handling data changes in connected React components and less of a clear "problem" we are experiencing).

Reply:
Yep, those are similar thoughts I had. I'll aim for "do nothing" right now, and we'll keep this in the back of our minds when it comes up again.

Reply:
@marcaaron I had another idea here... Would using the context API help at all? Does context work on only a portion of the React tree? It makes sense that it would... This is what I'm thinking:

1. SidebarLinks creates a context that includes all the Onyx data needed for the LHN
2. OptionRowLHN accesses the data from that context when calling `getOptionData()`

That would:
1. Remove the need to have the `connect()` calls in `SidebarUtils`
2. Not require OptionRowLHN to subscribe directly to the Onyx keys, saving a lot of overhead

Reply:
I like the context idea! I think in a follow up issue we should try the following and compare the performance to this implementation.
1. Use `withOnyx` in `SidebarLinks` and pass the reports and reportActions to `OptionsListUtilsLHN.getOrderedReportIDs()`.
2. Your context idea.

If option 2 is significantly better than option 1 we should do that. If it's only slightly better than option 1 would be more simple and would align with other components.

Reply:
I guess I'm not sure what advantage context would have in this case vs. something like:

- Subscribe in SidebarLinks to all the keys we need
- Pass the data down to the list for the renderItem() to use

We could set up a context provider in SidebarLinks and then have the individual option rows be consumers, but the effect would be similar since when the provider updates all the consumers will re-render. Did you have something else in mind?

Reply:
No, that's basically what I had in mind, but I think the difference is that instead of passing 32,000 copies of data (8 onyx keys to 4000 option rows), each row would only be accessing the data from context as a consumer when it renders. I imagine that's going to be a win for memory, but is 100% a hunch.

I like Neil's idea of going with what we have for now, then testing and profiling (both CPU and memory) the different alternatives.

Reply:
Ah, that's true. I wasn't thinking through that all the way.

Reply:
That looks so similar to what we were doing before though, gathering all the data for 4000 options at once and then passing it to each component. Like, were does `createOption()` exist in that flow? We'd still want it to be in `renderItem()` so that the heavy logic isn't done until render time.

Reply:
Similar but the key differences is that we were calculating all of the options before regardless of whether they were rendered. 

Your changes make it so that we only calculate a few options. 

So, just imagine that in the example I'm sharing all the data we are subscribing to via Onyx.connect() in SidebarUtils is just passed to the OptionsList as a prop and it can use them to create the 10 options that are rendered instead of 4000.